File: packages/transport/src/agent.ts
/**
 * Agent methods — methods exposed by the in-container `nexal-agent`.
 *
 * Frontend invokes them by wrapping in `agent/invoke`.
 * Used by `GatewayClient.invokeAgent`.
 */

export interface InitializeParams {
	client_name: string;
}
export interface InitializeResponse {
	default_shell?: string;
	cwd?: string;
}

export type InitializedParams = Record<string, never> | null | undefined;
export type InitializedResponse = null;

export type StreamKind = "stdout" | "stderr" | "pty";

export interface ProcessStartParams {
	process_id: string;
	argv: string[];
	cwd: string;
	env: Record<string, string>;
	tty: boolean;
	arg0: string | null;
	output_bytes_cap?: number;
}
export interface ProcessStartResponse {
	process_id: string;
}

export interface ProcessReadParams {
	process_id: string;
	after_seq: number;
	max_bytes: number;
	wait_ms: number;
}
export interface ProcessChunk {
	seq: number;
	stream: StreamKind;
	chunk: Uint8Array;
}
export interface ProcessReadResponse {
	chunks: ProcessChunk[];
	next_seq: number;
	exited: boolean;
	exit_code: number | null;
	closed: boolean;
	failure: string | null;
}

export interface ProcessTerminateParams {
	process_id: string;
}
export interface ProcessTerminateResponse {
	running: boolean;
}

export interface ProcessWriteParams {
	process_id: string;
	chunk: Uint8Array;
}
export interface ProcessWriteResponse {
	status: "accepted" | "unknown_process" | "stdin_closed" | "starting";
}

export interface FsReadFileParams {
	path: string;
}
export interface FsReadFileResponse {
	data: Uint8Array;
}

export interface FsWriteFileParams {
	path: string;
	data: Uint8Array;
}
export type FsWriteFileResponse = Record<string, never>;

export interface FsCreateDirectoryParams {
	path: string;
	recursive?: boolean;
}
export type FsCreateDirectoryResponse = Record<string, never>;

export interface FsGetMetadataParams {
	path: string;
}
export interface FsGetMetadataResponse {
	isDirectory: boolean;
	isFile: boolean;
	createdAtMs: number;
	modifiedAtMs: number;
}

export interface FsReadDirectoryParams {
	path: string;
}
export interface FsReadDirectoryEntry {
	fileName: string;
	isDirectory: boolean;
	isFile: boolean;
}
export interface FsReadDirectoryResponse {
	entries: FsReadDirectoryEntry[];
}

export interface FsRemoveParams {
	path: string;
	recursive?: boolean;
	force?: boolean;
}
export type FsRemoveResponse = Record<string, never>;

export interface FsCopyParams {
	sourcePath: string;
	destinationPath: string;
	recursive?: boolean;
}
export type FsCopyResponse = Record<string, never>;

export interface ProxyRegisterParams {
	socket_path: string;
	upstream_url: string;
	headers?: Record<string, string>;
}
export interface ProxyRegisterResponse {
	ok: boolean;
}

export interface ProxyUnregisterParams {
	socket_path: string;
}
export interface ProxyUnregisterResponse {
	ok: boolean;
}

/** Discriminated map used by `GatewayClient.invokeAgent` for type inference. */
export interface AgentMethods {
	initialize: { params: InitializeParams; result: InitializeResponse };
	initialized: { params: InitializedParams; result: InitializedResponse };
	"process/start": { params: ProcessStartParams; result: ProcessStartResponse };
	"process/read": { params: ProcessReadParams; result: ProcessReadResponse };
	"process/terminate": { params: ProcessTerminateParams; result: ProcessTerminateResponse };
	"process/write": { params: ProcessWriteParams; result: ProcessWriteResponse };
	"fs/read_file": { params: FsReadFileParams; result: FsReadFileResponse };
	"fs/write_file": { params: FsWriteFileParams; result: FsWriteFileResponse };
	"fs/create_directory": { params: FsCreateDirectoryParams; result: FsCreateDirectoryResponse };
	"fs/get_metadata": { params: FsGetMetadataParams; result: FsGetMetadataResponse };
	"fs/read_directory": { params: FsReadDirectoryParams; result: FsReadDirectoryResponse };
	"fs/remove": { params: FsRemoveParams; result: FsRemoveResponse };
	"fs/copy": { params: FsCopyParams; result: FsCopyResponse };
	"proxy/register": { params: ProxyRegisterParams; result: ProxyRegisterResponse };
	"proxy/unregister": { params: ProxyUnregisterParams; result: ProxyUnregisterResponse };
}

File: crates/utils/transport/src/agent.rs
use std::collections::HashMap;
use std::path::PathBuf;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct InitializeParams {
    pub client_name: String,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct InitializeResponse {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub default_shell: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cwd: Option<PathBuf>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProcessStartParams {
    pub process_id: String,
    pub argv: Vec<String>,
    pub cwd: PathBuf,
    pub env: HashMap<String, String>,
    pub tty: bool,
    pub arg0: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub output_bytes_cap: Option<usize>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProcessStartResponse {
    pub process_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProcessReadParams {
    pub process_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub after_seq: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_bytes: Option<usize>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub wait_ms: Option<u64>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StreamKind {
    Stdout,
    Stderr,
    Pty,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProcessChunk {
    pub seq: u64,
    pub stream: StreamKind,
    #[serde(with = "serde_bytes")]
    pub chunk: Vec<u8>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProcessReadResponse {
    pub chunks: Vec<ProcessChunk>,
    pub next_seq: u64,
    pub exited: bool,
    pub exit_code: Option<i32>,
    pub closed: bool,
    pub failure: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProcessWriteParams {
    pub process_id: String,
    #[serde(with = "serde_bytes")]
    pub chunk: Vec<u8>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WriteStatus {
    Accepted,
    UnknownProcess,
    StdinClosed,
    Starting,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProcessWriteResponse {
    pub status: WriteStatus,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProcessTerminateParams {
    pub process_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProcessTerminateResponse {
    pub running: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsReadFileParams {
    pub path: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsReadFileResponse {
    #[serde(with = "serde_bytes")]
    pub data: Vec<u8>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsWriteFileParams {
    pub path: String,
    #[serde(with = "serde_bytes")]
    pub data: Vec<u8>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsWriteFileResponse {}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsCreateDirectoryParams {
    pub path: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub recursive: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsCreateDirectoryResponse {}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsGetMetadataParams {
    pub path: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsGetMetadataResponse {
    pub is_directory: bool,
    pub is_file: bool,
    pub created_at_ms: i64,
    pub modified_at_ms: i64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsReadDirectoryParams {
    pub path: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsReadDirectoryEntry {
    pub file_name: String,
    pub is_directory: bool,
    pub is_file: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsReadDirectoryResponse {
    pub entries: Vec<FsReadDirectoryEntry>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsRemoveParams {
    pub path: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub recursive: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub force: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsRemoveResponse {}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsCopyParams {
    pub source_path: String,
    pub destination_path: String,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub recursive: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsCopyResponse {}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProxyRegisterParams {
    pub socket_path: String,
    pub upstream_url: String,
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub headers: HashMap<String, String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProxyRegisterResponse {
    pub ok: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProxyUnregisterParams {
    pub socket_path: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ProxyUnregisterResponse {
    pub ok: bool,
}

define_method_set! {
    /// Agent-side RPC method set.
    enum AgentMethod {
        Initialize        = "initialize"          => AgentInitialize(InitializeParams) -> InitializeResponse,
        Initialized       = "initialized"         => AgentInitialized(()) -> (),
        ProcessStart      = "process/start"       => AgentProcessStart(ProcessStartParams) -> ProcessStartResponse,
        ProcessRead       = "process/read"        => AgentProcessRead(ProcessReadParams) -> ProcessReadResponse,
        ProcessWrite      = "process/write"       => AgentProcessWrite(ProcessWriteParams) -> ProcessWriteResponse,
        ProcessTerminate  = "process/terminate"   => AgentProcessTerminate(ProcessTerminateParams) -> ProcessTerminateResponse,
        FsReadFile        = "fs/read_file"        => AgentFsReadFile(FsReadFileParams) -> FsReadFileResponse,
        FsWriteFile       = "fs/write_file"       => AgentFsWriteFile(FsWriteFileParams) -> FsWriteFileResponse,
        FsCreateDirectory = "fs/create_directory" => AgentFsCreateDirectory(FsCreateDirectoryParams) -> FsCreateDirectoryResponse,
        FsGetMetadata     = "fs/get_metadata"     => AgentFsGetMetadata(FsGetMetadataParams) -> FsGetMetadataResponse,
        FsReadDirectory   = "fs/read_directory"   => AgentFsReadDirectory(FsReadDirectoryParams) -> FsReadDirectoryResponse,
        FsRemove          = "fs/remove"           => AgentFsRemove(FsRemoveParams) -> FsRemoveResponse,
        FsCopy            = "fs/copy"             => AgentFsCopy(FsCopyParams) -> FsCopyResponse,
        ProxyRegister     = "proxy/register"      => AgentProxyRegister(ProxyRegisterParams) -> ProxyRegisterResponse,
        ProxyUnregister   = "proxy/unregister"    => AgentProxyUnregister(ProxyUnregisterParams) -> ProxyUnregisterResponse,
    }
}
